thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java

```diff
@@ -32,36 +39,261 @@
+                }
+            }
+        }
+    }
+
     private String generateId(@NonNull final TiPresenter presenter) {
         return presenter.getClass().getSimpleName()
                 + ":" + presenter.hashCode()
                 + ":" + System.nanoTime();
     }
+
+    @NonNull
+    private Application.ActivityLifecycleCallbacks getActivityLifecycleCallbacks() {
+        return new Application.ActivityLifecycleCallbacks() {
+            @Override
+            public void onActivityCreated(final Activity activity,
+                    final Bundle savedInstanceState) {
+                detectNewActivity(activity, savedInstanceState);
+            }
+
+            @Override
+            public void onActivityDestroyed(final Activity activity) {
+
+                TiLog.v(TAG, "destroying " + activity);
+                TiLog.v(TAG, "isFinishing = " + activity.isFinishing());
+                TiLog.v(TAG,
+                        "isChangingConfigurations = " + activity.isChangingConfigurations());
+
+                if (activity.isFinishing() && !activity.isChangingConfigurations()) {
+                    // detected Activity finish, no new Activity instance will be created
+                    // with savedInstanceState, clear saved presenters
+                    cleanupAfterFinish(activity);
+                } else {
+                    // don't leak old activity instances
+                    // scopeId is saved in savedInstanceState of finishing Activity.
+                    mScopeIdForActivity.remove(activity);
+                }
+            }
+
+            @Override
+            public void onActivityPaused(final Activity activity) {
+
+            }
+
+            @Override
+            public void onActivityResumed(final Activity activity) {
+
+            }
+
+            @Override
+            public void onActivitySaveInstanceState(final Activity activity,
+                    final Bundle outState) {
+                saveScopeId(activity, outState);
+            }
+
+            @Override
+            public void onActivityStarted(final Activity activity) {
+
+            }
+
+            @Override
+            public void onActivityStopped(final Activity activity) {
+
+            }
+        };
+    }
+
+    public void saveScopeId(final Activity activity, final Bundle outState) {
+        String id = mScopeIdForActivity.get(activity);
+        if (id == null) {
+            id = UUID.randomUUID().toString();
+            mScopeIdForActivity.put(activity, id);
+        }
+        outState.putString(TI_ACTIVITY_PRESENTER_SCOPE_KEY, id);
+    }
+
+    private synchronized ActivityScopedPresenters getScope(final Activity activity) {
+        final String id = mScopeIdForActivity.get(activity);
+        if (id != null) {
+            return mScopes.get(id);
+        }
+
+        return null;
+    }
+
+    private synchronized ActivityScopedPresenters getScopeOrCreate(final Activity activity) {
```

Comment:
Please use `getScope()` here as well and/or merge these two methods anyhow... They do the same thing